### ISSUE TITLE
Phi decoding Implementation

### ISSUE THREAD

**Issue by nqchieutb01** [yayayacc/MUR#3]
Thanks for your interesting work!
I have one question regarding clustering algorithm in Phi decoding. 
In file phi_decoding-mur.py line 71:
X = TfidfVectorizer().fit_transform(foresight_texts)
Why you use foresight_texts? Can you explain more about that, is this make sense when just have several candidates?
Also, I don't see you implement advantage function, is advantage function not important module?

Thank you so much for your time.

**Comment by yayayacc** [comment]
Hi, there!
Thank you so much for your attention.

- Q: About foresight_texts.
- A: We follow the idea of foresight sampling proposed in phi-decoding to use the foresight texts. You can refer to the original paper of phi-decoding for more details: https://arxiv.org/abs/2503.13288

- Q: Can you explain more about that, is this make sense when just have several candidates?
- A: Even with several candidates, we can still derive the larger cluster, which is a kind of self-consistency. According to the experiments in phi-decoding paper (Figure 3), it still works well under different FLOPs settings, even for relative small FLOPs (several candidates). 

- Q: Also, I don't see you implement advantage function, is advantage function not important module?
- A: In phi-decoding, the calculation of advantage is implemented by (foresight score of step t - foresight score of step t-1). However, as explained in MUR, we do not need foresight at each step. This foresight score is not available at each step in MUR, thus we do not include it.

Please feel free to discuss if you have further questions~

**Comment by nqchieutb01** [comment]
Thank you very much, I have another question about step-level uncertainty.

In your implementation:
cur_logp = output.cumulative_logprob / (len(output.token_ids) + 1e-8)

But in the paper there is a different version.
<img width="428" height="85" alt="Image" src="https://github.com/user-attachments/assets/ffcfbf28-12bd-4a6b-b5e8-3ca50b230688" />

Can you tell me the reason about this difference. 
cur_logp should be: 
cur_logp = (-1) * output.cumulative_logprob / (len(output.token_ids) + 1e-8)
Thank you!

**Comment by yayayacc** [comment]
Thank you for your meticulous attention!

You are right that the implementation is different.  However, I should clear that the different implementations are equivalent in math, and here is the proof:

Let me define different variables: 
1. m: the step level uncertainty in paper
2. M: the momentum uncertainty in paper
3. m_code(which is cur_logp as you metioned): the implementation in code:  output.cumulative_logprob / (len(output.token_ids) + 1e-8)
4. M_code(which we define "momentum_uncertainty" ***in code***): the implementation in code:  momentum_uncertainty = args.momentum_rate * momentum_uncertainty + (1 - args.momentum_rate) * cur_signal

We can infer that m_code = -m, M_code = -M.

What's more, the implementation of boundary condition for scaling ***in code*** is :
      np.exp(cur_signal) < np.exp(momentum_uncertainty) * args.scaling_rate
   => np.exp(m_code) < np.exp(M_code)* args.scaling_rate
   => np.exp(-m) < np.exp(-M)*args.scaling_rate
   => np.exp(M) / args.scaling_rate < np.exp(m)

And the boundary condition ***in paper*** is: np.exp(m) > np.exp(M) / args.scaling_rate. Although the implementation seems different, the resulted boundary conditions are equivalent in math.

Please feel free to discuss if you have further questions~

**Comment by nqchieutb01** [comment]
Oke, I get it, thank you very much :))
